rabbitmq/rabbitmq-stream-dotnet-client
# Reliable  Producer

Reliable  Producer 
--
part of https://github.com/rabbitmq/rabbitmq-stream-dotnet-client/issues/92

RProducer:
-  is able to reconnect and invalidate old messages.
- trace the sent and confirmed messages in an internal `waitForConfirmation` list (as the go/java clients do) ex:
- invalidate the messages without confirmation ( in 2 seconds ) 

```csharp
 var p = await ReliableProducer.CreateRProducer(new ReliableProducerConfig()
            {
                StreamSystem = system,
                Stream = stream,
                ConfirmationHandler = confirmation =>
                {
                    if (confirmation.Status == ConfirmationStatus.Confirmed)
                    {
 
                        countConf++;
                    }
                    else
                    {
                        countError++;
                    }
           
```

Reconnection
--

By default it uses an `BackOffReconnectStrategy` to reconnect the client, but the user can implement another behaviour just implement the `IReconnectStrategy` interface.
```csharp
var p = await ReliableProducer.CreateReliableProducer(new ReliableProducerConfig()
            {
                ReconnectStrategy = myReconnectStrategy            
```




How to test:
```csharp 

public async Task TestRProducer()
    {
        var config = new StreamSystemConfig
        {
        
        };
        const string stream = "mystream-r";
        var system = await StreamSystem.Create(config);
        try
        {
            await system.DeleteStream(stream);
        }
        catch (Exception e)
        {
            // ignored 
        }

        await system.CreateStream(new StreamSpec(stream)
        {
            MaxLengthBytes = 2073741824
        });
        var countConf = 0;
        var countError = 0;
        var gc = 0;
        const int totalMessages = 10_000_000;
        var run = Task.Run(async () =>
        {
            var p = await ReliableProducer.CreateReliableProducer(new ReliableProducerConfig()
            {
                StreamSystem = system,
                Stream = stream,
                Reference = "myref",
                ConfirmationHandler = confirmation =>
                {
                    
                    gc++;
                   if (confirmation.Status == ConfirmationStatus.Confirmed)
                    {
 
                        countConf++;
                    }
                    else
                    {
                        countError++;
                    }

                    if ( gc % 50_000 == 0)
                    {
                        Console.WriteLine(
                            $"-Partial confirmed: {countConf} - TimeOutError: {countError} - total: {countError + countConf}");
                    }
                    
                    if ( gc  == totalMessages)
                    {
                        Console.WriteLine(
                            $"-END confirmed: {countConf} - TimeOutError: {countError} - total: {countError + countConf}");
                    }

                    return Task.CompletedTask;
                }
            });
            var start = DateTime.Now;
            for (var i = 0; i < totalMessages; i++)
            {
                if ((i +1) % 100_000 == 0)
                {
                    Console.WriteLine($"sent {i + 1}");
                }

                try
                {
                    await p.Send(new Message(Encoding.UTF8.GetBytes($"hello {i}")));
                }
                catch (Exception e)
                {
                    Console.WriteLine($"Error {e}");
                }
            }

            Console.WriteLine($"End...Done {DateTime.Now - start}");
            Thread.Sleep(5000);
            // await p.Close();
        });


        for (var i = 0; i < 50; i++)
        {
            Thread.Sleep(5000);
            try
            {
                Console.WriteLine($"Killing all the connections...");
                var r = HttpKillConnections();
                r.Wait();
            }
            catch (Exception e)
            {
                Console.WriteLine($"fail to kill {e}");
            }
        }

        run.Wait();
        Thread.Sleep(50000);
        Console.WriteLine("Done");
    }

```

Handle Metadata update:
---
`MetadataUpdate` is raised when the stream is deleted or changes the topology.

The `ReliableProducer` is able to:
- Detect the event 
- reconnect the client in case the stream still exists 
- close the producer and the TCP connection if the stream does/t exist anymore.

See [here](https://github.com/rabbitmq/rabbitmq-stream-dotnet-client/pull/104/files#diff-4d1f7a61aec33fb380769c1afb8a426357ef3c7983db71093698692941c1e078R146-R168)

How to test:

See the [HandleDeleteStreamWithMetaDataUpdate ](https://github.com/rabbitmq/rabbitmq-stream-dotnet-client/pull/104/files#diff-12c164058772f192c105f353c1c994ac1bc35825832773dce61c95ffa44b67dcR173-R194)

it is enough to remove a stream when the producer is running.

See also the [ HandleChangeStreamConfigurationWithMetaDataUpdate](https://github.com/rabbitmq/rabbitmq-stream-dotnet-client/pull/104/files#diff-12c164058772f192c105f353c1c994ac1bc35825832773dce61c95ffa44b67dcR198-R221) 

Auto Generate Sequence
---
ReliableProducer Automatically retrieves the last stored PublishingID and It continues from there.
